src/Common/src/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -393,6 +396,181 @@ private static ComputedInstanceFieldLayout ComputeSequentialFieldLayout(Metadata
             return computedLayout;
         }
 
+        protected virtual ComputedInstanceFieldLayout ComputeAutoFieldLayout(MetadataType type, int numInstanceFields)
+        {
+            // For types inheriting from another type, field offsets continue on from where they left off
+            LayoutInt cumulativeInstanceFieldPos = ComputeBytesUsedInParentType(type);
+
+            var layoutMetadata = type.GetClassLayout();
+
+            int packingSize = ComputePackingSize(type, layoutMetadata);
+            LayoutInt largestAlignmentRequired = LayoutInt.One;
+
+            var offsets = new FieldAndOffset[numInstanceFields];
+            int fieldOrdinal = 0;
+
+            // Initialize three different sets of lists to hold the instance fields
+            //   1. List of value class fields
+            //   2. Lists of GC Pointer fields. These will be stored in a dictionary indexed by the size of the field
+            //   3. Lists of remaining fields. These will be stored in a dictionary indexed by the size of the field
+            List<FieldDesc> instanceValueClassFieldDescs = new List<FieldDesc>();
+            Dictionary<int, List<FieldDesc>> instanceGCPointerFieldsBySize = new Dictionary<int, List<FieldDesc>>();
```

Comment:
The CoreCLR implementation of this algorithm achieves this with a couple simple arrays. No hashtables, etc.

This implementation feels much more heavyweigth/slower.

Reply:
Yes it definitely is...I'll spend some time to determine how I can limit/eliminate the use of hash tables, though this will come at the cost of more time looping over the fields to gather the appropriate data. The CoreCLR implementation takes advantage of the fact that several of the constructs have already been prepared.

Reply:
Are there any assumptions/constant values I could take advantage of to improve this? For instance, in CoreCLR the largest primitive size `MAX_LOG2_PRIMITIVE_FIELD_SIZE` is used to set the upper bound on field sizes to expect. Is there a similar construct I could use here?

Reply:
+1 to what Jan said. This code is going to run as part of the runtime type loader too because the type system is shared. If we can get away with just arrays, we should. Using fancy data structures from BCL could cause nasty recursion (e.g. we're in the middle of building an eetype for Dictionary when we find out we actually need the eetype for Dictionary to do the rest of the building)

Reply:
I edited the implementation to only use Arrays. Can you let me know what you think of the changes and any further improvements you'd like to see?